React Cloud docs (#17020)
ckeditor/ckeditor5: docs/getting-started/integrations-cdn/react-default-cdn.md

{@snippet installation/integrations/framework-integration}

> # React rich text editor component in CDN

Reviewer: I truly despise the " (npm)" approach to be honest. But I have next to zero emotional ties to these titles. Possible relinking incoming.

<p>
	<a href="https://www.npmjs.com/package/@ckeditor/ckeditor5-react" target="_blank" rel="noopener">
		<img src="https://badge.fury.io/js/%40ckeditor%2Fckeditor5-react.svg" alt="npm version" loading="lazy">
	</a>
</p>

> This guide explains how to integrate CKEditor5 into your React application using CDN.

Reviewer 2: ```suggestion
This guide explains how to integrate CKEditor&nbsp;5 into your React application using CDN.
```

### Setting up the project

This guide assumes you have a ready React project. If not, you can create a basic React project using [Vite](https://vitejs.dev/). Refer to the [React documentation](https://react.dev/learn/start-a-new-react-project) to learn how to set up a project in the framework.

> ### Installing from npm

Author: @Mati365 is it valid?

> ### Installing from npm

Reviewer 3: @illia-stv Tricky question. React integration is installed from NPM, but it can use `useCKEditorCloud` which uses CDN under the hood. I'd keep `NPM` because we describe `@ckeditor/ckeditor5-react` here. imho

> ### Installing from npm

Author: I would agree with that, lets stay with `NPM` then.

> ### Installing from npm

Reviewer 4: Let's make it more specific, like Install React component package or something like that

> ### Installing from npm

Reviewer 5: I'm fine with npm, then.

Here is an example:

```js
import React from 'react';
import { CKEditor, useCKEditorCloud } from '@ckeditor/ckeditor5-react';

const CKEditorDemo = () => {
	const cloud = useCKEditorCloud( {
		version: '43.0.0',
		plugins: {
			// New one, example: local UMD plugins
			Plugin1: async () => import( './your-local-import.umd.js' ),

			// New one, example: local external imports
			// Additional bundler configuration must be used, like this one:
			// https://github.com/crcong/vite-plugin-externals
			Plugin11: async () => import( './your-local-import' ) as unknown as Promise<Your_Type>,

			// New one, example: CDN 3rd party
			Plugin2: [
				'https://cdn.example.com/plugin2.js',
				'https://cdn.example.com/plugin2.css'
			],

			// Old one, verbose for more advanced plugins. It's still available.
			Plugin3: {
				scripts: [ 'https://cdn.example.com/plugin3.js' ],
				stylesheets: [ 'https://cdn.example.com/plugin3.css' ],

				// Optional, if it's not passed then the type of `Plugin3` will be picked from `Window`
				checkPluginLoaded: () => ( window as any ).Plugin3

> 			}

Reviewer 3: That `new one`, `old one` and `it's still availble` comments should be removed. It was used only for demo purposes

		}
	} );

	if ( cloud.status === 'error' ) {
		return <div>Error!</div>;
	}

	if ( cloud.status === 'loading' ) {
		return <div>Loading...</div>;
	}

> 	const { Plugin1, Plugin2, ... } = cloud.loadedPlugins;

Reviewer 5: Tip for all of us: let's try to make snippets compilable/parsable.
```suggestion
	const { Plugin1, Plugin2, /* More plugins... */ } = cloud.loadedPlugins;
```

This guide assumes you have a React project. You can create a basic React project using [Vite](https://vitejs.dev/). Refer to the [React documentation](https://react.dev/learn/start-a-new-react-project) to learn how to set up a project in the framework.

> ### Installing react component from npm

Reviewer 6: ```suggestion
### Installing React component from npm
```

Install the `@ckeditor/ckeditor5-react` package:

```bash
npm install @ckeditor/ckeditor5-react
```

The `useCKEditorCloud` hook is responsible for returning information that:

> * CKEditor is still being downloaded from the CDN with status = "loading".

Reviewer: Don't we want inline code here for statuses?

> * CKEditor is still being downloaded from the CDN with status = "loading".

Reviewer: Why did we stop using the proper "CKEditor 5" name, by the way?

- An error occurred during the download when status = "error", then further information is returned in the error field.
- Returning the editor in the data field and its dependencies when status = "success".

Use the `<CKEditor>` component inside your project. The below example shows how to use the component with the open-source plugins.

```js
import React from "react";
import { CKEditor, useCKEditorCloud } from "@ckeditor/ckeditor5-react";

const CKEditorDemo = () => {
	const cloud = useCKEditorCloud( {

> 		version: "43.0.0",

Reviewer 2: Maybe better to use current release version, WDYT?

```suggestion
		version: "{@var ckeditor5-version}",
```

Reviewer: I'm all in for that solution, we won't need to update it manually.

Reviewer 5: I wondered if the variable would work in snippets. And yes - it does. So, it's a good idea.

	} );

	if ( cloud.status === "error" ) {
		return <div>Error!</div>;
	}

	if ( cloud.status === "loading" ) {
		return <div>Loading...</div>;
	}

	const {
		ClassicEditor,
		Bold,
		Essentials,
		Italic,
		Paragraph,
		Undo,
	} = cloud.CKEditor;

	return (
		<CKEditor
			editor={ ClassicEditor }

> 			data={ "<p>Hello world!!!</p>" }

Reviewer 6: ```suggestion
			data={ "<p>Hello world!</p>" }
```

We don't need to be so passive-aggressive.

			config={{
			toolbar: {
				items: ["undo", "redo", "|", "bold", "italic"],
			},
			plugins: [ Bold, Essentials, Italic, Paragraph, Undo ],
			}}
		/>
	);
};
```

To use premium plugins, set the `premium` property to `true` in the `useCKEditorCloud` configuration and provide your license key in the `CKEditor` configuration.

```js
import React from "react";
import { CKEditor, useCKEditorCloud } from "@ckeditor/ckeditor5-react";

const CKEditorDemo = () => {
	const cloud = useCKEditorCloud({

> 		version: "43.0.0",

Reviewer 2: Maybe better to use current release version, WDYT?

```suggestion
		version: "{@var ckeditor5-version}",
```

		premium: true,
	} );

	if ( cloud.status === "error" ) {
		return <div>Error!</div>;
	}

	if ( cloud.status === "loading" ) {
		return <div>Loading...</div>;
	}

	const {
		ClassicEditor,
		Bold,
		Essentials,
		Italic,
		Mention,
		Paragraph,
		Undo,
	} = cloud.CKEditor;

	const { SlashCommand } = cloud.CKEditorPremiumFeatures!;

	return (
		<CKEditor
			editor={ ClassicEditor }

> 			data={ "<p>Hello world!!!</p>" }

Reviewer 6: ```suggestion
			data={ "<p>Hello world!</p>" }
```

We don't need to be so passive-aggressive.

Reviewer 5: I would say that the creator of the snippet was more excited. But, yeah, right, hah

			config={{
				licenseKey: "<YOUR_LICENSE_KEY>",
				toolbar: {
					items: [ "undo", "redo", "|", "bold", "italic" ],
				},
				plugins: [
					Bold,
					Essentials,
					Italic,
					Mention,
					Paragraph,
					Undo,
					SlashCommand,
				],
			}}
		/>
	);
};
```

### Usage with CKBox

To use `CKBox`, specify the version and theme (optionally) in the `useCKEditorCloud` configuration. Also, remember about the actual plugin configuration inside `<CKEditor/>` component.

```js
import React from "react";
import { CKEditor, useCKEditorCloud } from "@ckeditor/ckeditor5-react";

const CKEditorDemo = () => {
	const cloud = useCKEditorCloud({
	version: "43.0.0",
	ckbox: {

> 		version: "2.5.1",

Reviewer 2: Will `"latest"` be also a proper value? :thinking:

Reviewer 3: At this moment - no. Imho, latest should be avoided to protect integration against breaking changes.

			// Optional - it's already 'lark' by default.
			theme: "lark",
		},
	} );

	if ( cloud.status === "error" ) {
		return <div>Error!</div>;
	}

	if ( cloud.status === "loading" ) {
		return <div>Loading...</div>;
	}

	const {
		ClassicEditor,
		Bold,
		Essentials,
		Italic,
		Mention,
		Paragraph,
		Undo,
		CKBox,
		CKBoxImageEdit,
	} = cloud.CKEditor;

	return (
		<CKEditor
			editor={ ClassicEditor }

> 			data={ "<p>Hello world!!!</p>" }

Reviewer 6: ```suggestion
			data={ "<p>Hello world!</p>" }
```

We don't need to be so passive-aggressive.

			config={ {
				toolbar: {
					items: [ "undo", "redo", "|", "bold", "italic" ],
				},
				plugins: [
					Bold,
					Essentials,
					Italic,
					Mention,
					Paragraph,
					Undo,
					CKBox,
					CKBoxImageEdit,
				],
				ckbox: {
					tokenUrl: "https://api.ckbox.io/token/demo",
					forceDemoLabel: true,
					allowExternalImagesEditing: [ /^data:/, /^i.imgur.com\//, "origin" ],
				},
			} }
		/>
	);
};
```

### Usage with external plugins

There are various ways to use external plugins. Here is a list of them:

* **Local UMD Plugins:** Dynamically import local UMD modules using the `import()` syntax.
* **Local External Imports:** Load external plugins locally using additional bundler configurations (such as Vite).

> * **CDN 3rd Party Plugins:** Load JavaScript and CSS files from a CDN by specifying the URLs.

Reviewer 6: ```suggestion
* **CDN Third-Party Plugins:** Load JavaScript and CSS files from a CDN by specifying the URLs.
```

* **Verbose Configuration:** Advanced plugin loading with options to specify both script and style sheet URLs, along with an optional `checkPluginLoaded` function to verify the plugin has been correctly loaded into the global scope.

Here is an example:

```js
import React from 'react';
import { CKEditor, useCKEditorCloud } from '@ckeditor/ckeditor5-react';

const CKEditorDemo = () => {
	const cloud = useCKEditorCloud( {
		version: '{@var ckeditor5-version}',
		plugins: {

> 			PluginUMD: async () => import( './your-local-import.umd.js' ),
> 			PluginLocalImport: async () => import( './your-local-import' ),

Reviewer 6: ```suggestion
			PluginUMD: () => import( './your-local-import.umd.js' ),
			PluginLocalImport: () => import( './your-local-import' ),
```

Since there is no `await`, the `async` keyword should not be needed.

Reviewer 5: You're right. But shouldn't this be the other way around - shouldn't we add `await`? This dynamic import is asynchronous AFAIR.

Reviewer 3: tbh, I'd stay with `async` prefix without `await`. It's a little bit more verbose, but it's also more descriptive for the users which may be not super familiar with dynamic imports.

Reviewer 5: Is it recommended? Won't such a syntax trigger some linter rules? @filipsobol

Reviewer 6: It will trigger the [`require-await`](https://eslint.org/docs/latest/rules/require-await) rule if the end project has it enabled.

I suggested removing it only for technical correctness. I'm fine with keeping it if it makes things more clear. In this case, I don't think it's important for the end developer to know whether `import()` is async or not because it won't affect them in any way.

This guide assumes you have a React project. You can create a basic React project using [Vite](https://vitejs.dev/). Refer to the [React documentation](https://react.dev/learn/start-a-new-react-project) to learn how to set up a project in the framework.

### Installing from npm

Install the `@ckeditor/ckeditor5-react` package:

```bash
npm install @ckeditor/ckeditor5-react
```

Use the `<CKEditor>` component inside your project. The below example shows how to use the component with the open-source plugins.

> 

Reviewer 4: I miss the info what `useCKEditorCloud` does (adding all needed scripts to a page).

```js
import React from "react";
import { CKEditor, useCKEditorCloud } from "@ckeditor/ckeditor5-react";

const CKEditorDemo = () => {
	const cloud = useCKEditorCloud( {
		version: "43.0.0",
	} );

	if ( cloud.status === "error" ) {
		return <div>Error!</div>;
	}

	if ( cloud.status === "loading" ) {
		eturn <div>Loading...</div>;
	}

	const {

> 		ClassicEditor: ClassicEditorBase,

Reviewer 4: `ClassicEditorBase` not needed

		premium: true,
	});

	if ( cloud.status === "error" ) {
		return <div>Error!</div>;
	}

	if ( cloud.status === "loading" ) {
		return <div>Loading...</div>;
	}

	const {
		ClassicEditor: ClassicEditorBase,
		Bold,
		Essentials,
		Italic,
		Mention,
		Paragraph,
		Undo,
	} = cloud.CKEditor;

	const { SlashCommand } = cloud.CKEditorPremiumFeatures!;

	return (
		<CKEditor
			editor={ ClassicEditorBase }
			data={ "<p>Hello world!!!</p>" }
			config={{
				licenseKey: "<YOUR_LICENSE_KEY>",
				toolbar: {
					items: [ "undo", "redo", "|", "bold", "italic" ],
				},
				plugins: [
					Bold,
					Essentials,
					Italic,
					Mention,
					Paragraph,
					Undo,
					SlashCommand,
				],
			}}
		/>
	);
};
```

### Usage with CKBox

To use `CKBox`, specify the version and theme (optionally) in the `useCKEditorCloud` configuration. Also, remember about the actual plugin configuration inside `<CKEditor/>` component.

```js
import React from "react";
import { CKEditor, useCKEditorCloud } from "@ckeditor/ckeditor5-react";

const CKEditorDemo = () => {
	const cloud = useCKEditorCloud({
	version: "43.0.0",
	ckbox: {
		version: "2.5.1",
		// Optional - it's already 'lark' by default.

> 		theme: "lark",

Reviewer 4: not sure what we mean by optional, if it's already loads lark, then it's not needed here.

Reviewer: We may, however, want to keep it to show having the parameter.

				},
				plugins: [
					Bold,
					Essentials,
					Italic,
					Mention,
					Paragraph,
					Undo,
					CKBox,
					CKBoxImageEdit,
				],
				ckbox: {
					tokenUrl: "https://api.ckbox.io/token/demo",
					forceDemoLabel: true,
					allowExternalImagesEditing: [/^data:/, /^i.imgur.com\//, "origin"],
				},
			}}
		/>
	);
};
```

### Usage with external plugins

There are various ways to use external plugins. Here is a list of them:

- **Local UMD Plugins:** Dynamically import local UMD modules using the `import()` syntax.
- **Local External Imports:** Load external plugins locally using additional bundler configurations (e.g., Vite).
- **CDN 3rd Party Plugins:** Load JavaScript and CSS files from a CDN by specifying the URLs.
- **Verbose Configuration:** Advanced plugin loading with options to specify both script and stylesheet URLs, along with an optional `checkPluginLoaded` function to verify the plugin has been correctly loaded into the global scope.

Here is an example:

```js
import React from 'react';
import { CKEditor, useCKEditorCloud } from '@ckeditor/ckeditor5-react';

const CKEditorDemo = () => {
	const cloud = useCKEditorCloud( {
		version: '43.0.0',
		plugins: {

> 			// New one, example: local UMD plugins

Reviewer 4: "New one" was an internal comment, this shouldn't appear in the docs.

		return <div>Error!</div>;
	}

	if ( cloud.status === 'loading' ) {
		return <div>Loading...</div>;
	}

	const {
		ClassicEditor,
		Bold,
		Essentials,
		Italic,
		Paragraph,
		Undo,
	} = cloud.CKEditor;

	return (
		<CKEditor
			editor={ ClassicEditor }
			data={ '<p>Hello world!</p>' }
			config={ {
				licenseKey: 'GPL',
				toolbar: {
					items: [ 'undo', 'redo', '|', 'bold', 'italic' ],
				},
				plugins: [ Bold, Essentials, Italic, Paragraph, Undo ],
			} }
		/>
	);
};
```

### Using the component with premium plugins

To use premium plugins, set the `premium` property to `true` in the `useCKEditorCloud` configuration and provide your license key in the `CKEditor` configuration.

```js
import React from 'react';
import { CKEditor, useCKEditorCloud } from '@ckeditor/ckeditor5-react';

const CKEditorDemo = () => {
	const cloud = useCKEditorCloud( {
		version: '{@var ckeditor5-version}',
		premium: true,
	} );

	if ( cloud.status === 'error' ) {
		return <div>Error!</div>;
	}

	if ( cloud.status === 'loading' ) {
		return <div>Loading...</div>;
	}

	const {
		ClassicEditor,
		Bold,
		Essentials,
		Italic,
		Mention,
		Paragraph,
		Undo,
	} = cloud.CKEditor;

> 	const { SlashCommand } = cloud.CKEditorPremiumFeatures!;

Reviewer 5: ```suggestion
	const { SlashCommand } = cloud.CKEditorPremiumFeatures;
```

1. An {@link module:utils/eventinfo~EventInfo `EventInfo`} object.
2. An {@link module:core/editor/editor~Editor `Editor`} instance.

The `<useCKEditorCloud>` component supports the following properties:

* `version` (required) &ndash; The version of CKEditor Cloud Services to use.

> * `languages` &ndash; The languages to load.

Reviewer 2: Is `languages` means `translations` in this case? If yes that `en` is the default language and don't need to be downloaded separately.

> * `languages` &ndash; The languages to load.

Reviewer 3: Yep /cc @illia-stv

> * `languages` &ndash; The languages to load.

Reviewer 3: @gorzelinski We can, but there is already `language` configuration option in editor, and `languages` are supposed to represent all possible values that can be passed to editor configuration `language` property. I'd keep it as it is.

* `premium` &ndash; If `true` then the premium features will be loaded.

> * `id` &ndash; The editor ID. When this property changes, the component restarts the editor with new data instead of setting it on an initialized editor.

Reviewer 3: This prop does not exist in hook config

* `onBlur` &ndash; A function called when the editor was blurred. See the {@link module:engine/view/document~Document#event:blur `editor.editing.view.document#blur`} event.
* `onFocus` &ndash; A function called when the editor was focused. See the {@link module:engine/view/document~Document#event:focus `editor.editing.view.document#focus`} event.
* `onError` &ndash; A function called when the editor has crashed during the initialization or during the runtime. It receives two arguments: the error instance and the error details. Error details is an object that contains two properties:

> 	* `{String} phase`: `'initialization'|'runtime'` &ndash; Informs when the error has occurred (during the editor or context initialization, or after the initialization).
> 	* `{Boolean} willEditorRestart` &ndash; When `true`, it means that the editor component will restart itself.

Reviewer 5: The info about types is useful but it's inconsistent with the bullet points above. Maybe we should remove it?
```suggestion
	* `phase`: `'initialization'|'runtime'` &ndash; Informs when the error has occurred (during the editor or context initialization, or after the initialization).
	* `willEditorRestart` &ndash; When `true`, it means that the editor component will restart itself.
```

* `id` &ndash; The context ID. When this property changes, the component restarts the context with its editor and reinitializes it based on the current configuration.
* `onReady` &ndash; A function called when the context is ready and all editors inside were initialized with the `context` instance. This callback is also called after the reinitialization of the component if an error has occurred.
* `onError` &ndash; A function called when the context has crashed during the initialization or during the runtime. It receives two arguments: the error instance and the error details. Error details is an object that contains two properties:

> 	* `{String} phase`: `'initialization'|'runtime'` &ndash; Informs when the error has occurred (during the editor or context initialization, or after the initialization).
> 	* `{Boolean} willContextRestart` &ndash; When `true`, it means that the context component will restart itself.

Reviewer 5: Similarly here.
```suggestion
	* `phase`: `'initialization'|'runtime'` &ndash; Informs when the error has occurred (during the editor or context initialization, or after the initialization).
	* `willContextRestart` &ndash; When `true`, it means that the context component will restart itself.
```

				return () => {
						setMounted( false );
				};
		}, [] );
 if ( cloud.status === 'error' ) {
		console.error( cloud );
		return <div>Error!</div>;
	}

	if ( cloud.status === 'loading' ) {
		return <div>Loading...</div>;
	}
		const { DecoupledEditor, Bold, Italic, Paragraph, Essentials } = cloud.CKEditor;
		return (
				<div>
						<div ref={ editorToolbarRef }></div>
						<div>
								{ isMounted && (
										<CKEditor
												editor={ DecoupledEditor }
												data='<p>Hello from CKEditor 5 decoupled editor!</p>'
												config={ {
														plugins: [ Bold, Italic, Paragraph, Essentials ],
														toolbar: [ 'undo', 'redo', '|', 'bold', 'italic' ]
												} }
												onReady={ ( editor ) => {
														if ( editorToolbarRef.current ) {
																editorToolbarRef.current.appendChild( editor.ui.view.toolbar.element );
														}
												}}
												onAfterDestroy={ ( editor ) => {
														if ( editorToolbarRef.current ) {
																Array.from( editorToolbarRef.current.children ).forEach( child => child.remove() );
														}
												}}
										/>
								) }
						</div>
				</div>

> 		);

Reviewer 5: The formatting/spacing of this snippet seems to be off.

CKEditor&nbsp;5 supports {@link getting-started/setup/ui-language multiple UI languages}, and so does the official React component. Follow the instructions below to translate CKEditor&nbsp;5 in your React application.

Pass the language you need into the `translations` array inside the configuration in the `useCKEditorCloud`.

```js
import React from "react";
import { CKEditor, useCKEditorCloud } from "@ckeditor/ckeditor5-react";

const CKEditorDemo = () => {

> 	const cloud = useCKEditorCloud({
> 		version: "43.0.0",
> 	translations: [ "de" ]
> 	});

Reviewer 5: ```suggestion
	const cloud = useCKEditorCloud({
		version: "43.0.0",
	        translations: [ "de" ]
	});
```

	if (cloud.status === "error") {
		return <div>Error!</div>;
	}

	if (cloud.status === "loading") {
		return <div>Loading...</div>;
	}

	const {
		ClassicEditor: ClassicEditorBase,
		Bold,
		Essentials,
		Italic,

> 		Mention,

Reviewer 5: We can apply the mention suggestions here as well.
```suggestion
```

		Paragraph,
		Undo,
	} = cloud.CKEditor;

	return (
		<CKEditor
			editor={ClassicEditorBase}
			data={"<p>Hello world!!!</p>"}
			config={{

> 				toolbar: {
> 					items: ["undo", "redo", "|", "bold", "italic"],
> 				},

Reviewer 5: ```suggestion
				toolbar: ["undo", "redo", "|", "bold", "italic"],
```

CKEditor&nbsp;5 supports {@link getting-started/setup/ui-language multiple UI languages}, and so does the official React component. Follow the instructions below to translate CKEditor&nbsp;5 in your React application.

Pass the language you need into the `translations` array inside the configuration in the `useCKEditorCloud`.

```js
import React from "react";
import { CKEditor, useCKEditorCloud } from "@ckeditor/ckeditor5-react";

const CKEditorDemo = () => {
	const cloud = useCKEditorCloud({
		version: "{@var ckeditor5-version}",
		languages: [ "de" ]
	});

	if ( cloud.status === "error" ) {
		return <div>Error!</div>;
	}

	if ( cloud.status === "loading" ) {
		return <div>Loading...</div>;
	}

	const {
		ClassicEditor: ClassicEditorBase,
		Bold,
		Essentials,
		Italic,
		Paragraph,
		Undo,
	} = cloud.CKEditor;

	return (
		<CKEditor
			editor={ ClassicEditorBase }

> 			data={ "<p>Hello world!!!</p>" }

Reviewer 6: ```suggestion
			data={ "<p>Hello world!</p>" }
```

We don't need to be so passive-aggressive.

CKEditor&nbsp;5 supports {@link getting-started/setup/ui-language multiple UI languages}, and so does the official React component. Follow the instructions below to translate CKEditor&nbsp;5 in your React application.

> Pass the language you need into the `translations` array inside the configuration in the `useCKEditorCloud`.

Reviewer 2: This is not reflected on the snippet below, should it be? :thinking: 

There is only `languages` key.

Reviewer 5: Good catch. I'm guessing it should be the `languages` key, right? @Mati365

Reviewer 4: Maybe it makes sense to rename it to fit editor config.

Reviewer 3: It's named language in editor config, https://ckeditor.com/docs/ckeditor5/latest/api/module_core_editor_editorconfig-EditorConfig.html#member-language . @gorzelinski yep, it should be languages

Reviewer 2: Setting the [language of the ui and/or content](https://ckeditor.com/docs/ckeditor5/latest/getting-started/setup/ui-language.html#setting-the-language-of-the-content) is a bit different thing than [loading the translations](https://ckeditor.com/docs/ckeditor5/latest/getting-started/setup/ui-language.html#loading-additional-languages-from-npm-or-cdn).

Reviewer 3: Renamed to translations
